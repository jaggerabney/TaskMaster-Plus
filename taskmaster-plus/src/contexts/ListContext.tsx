"use client";

import React, { createContext, useReducer } from "react";

export type List = {
  id: number;
  title: string;
  // user to come
  // tasks to come
};

export type ListContextType = {
  state: { lists: List[] };
  dispatch: React.Dispatch<ListAction>;
};

export enum ListActions {
  CREATE = "CREATE",
  READ = "READ",
  UPDATE = "UPDATE",
  DELETE = "DELETE"
}

interface ListAction {
  type: ListActions;
  payload: {
    id: number;
    title: string;
  };
}

interface ListState {
  lists: List[];
}

export const listReducer = (
  state: ListState,
  action: ListAction
): ListState => {
  switch (action.type) {
    case "CREATE":
      return {
        lists: [
          ...state.lists,
          { id: action.payload.id, title: action.payload.title }
        ]
      };
    default:
      return state;
  }
};

const ListContext = createContext<ListContextType>({
  state: { lists: [] },
  dispatch: () => null
});

const ListProvider = ({ children }: { children: React.ReactNode }) => {
  const [state, dispatch] = useReducer(listReducer, {
    lists: []
  });

  return (
    <ListContext.Provider value={{ state, dispatch }}>
      {children}
    </ListContext.Provider>
  );
};

export { ListProvider, ListContext };
