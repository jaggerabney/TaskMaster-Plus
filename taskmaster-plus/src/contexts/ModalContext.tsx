"use client";

import React, { createContext, useState } from "react";

export interface ModalContextType {
  modalIsOpen: boolean;
  openModal: () => void;
  closeModal: () => void;
}

export const ModalContext = createContext<ModalContextType>({
  modalIsOpen: false,
  openModal: () => {},
  closeModal: () => {}
});

export const ModalContextProvider = (props: React.PropsWithChildren) => {
  const [modalIsOpen, setModalIsOpen] = useState(false);

  function openModal() {
    setModalIsOpen(true);
  }

  function closeModal() {
    setModalIsOpen(false);
  }

  return (
    <ModalContext.Provider value={{ modalIsOpen, openModal, closeModal }}>
      {props.children}
    </ModalContext.Provider>
  );
};
