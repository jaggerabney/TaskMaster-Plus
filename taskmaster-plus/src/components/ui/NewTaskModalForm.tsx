"use client";

import React, { useContext, useRef } from "react";

import Button from "./Button";
import { ListContext, Task } from "@/contexts/ListContext";

export interface NewTaskModalFormType {
  onSubmit: (task: Task) => void;
  onCancel: () => void;
}

const NewTaskModalForm: React.FC<NewTaskModalFormType> = ({
  onSubmit,
  onCancel
}) => {
  const listContext = useContext(ListContext);
  const listDropdownRef = useRef<HTMLOptionElement>(null);

  function formSubmitHandler(event: React.FormEvent) {
    event.preventDefault();

    const eventTarget = event.target as typeof event.target & {
      title: { value: string };
      datetime: { value: Date };
      description: { value: string };
    };
    const id = Math.random();
    const listId = Number(listDropdownRef.current?.id);

    const newTask: Task = {
      id,
      listId,
      title: eventTarget.title.value || "Untitled task",
      completed: false,
      dueDate: new Date(eventTarget.datetime.value),
      description: eventTarget.description.value
    };

    onSubmit(newTask);
  }

  return (
    <form onSubmit={formSubmitHandler} className="flex flex-col gap-4">
      <div
        id="list-dropdown-wrapper"
        className="flex flex-row gap-4 items-center"
      >
        <label htmlFor="list" className="text-xl">
          List:
        </label>
        <select
          id="list"
          name="list"
          className="border border-black rounded text-xl p-2"
        >
          {listContext.state.lists.map((list) => (
            <option
              key={list.id}
              id={list.id.toString()}
              value={list.title}
              ref={listDropdownRef}
            >
              {list.title}
            </option>
          ))}
        </select>
      </div>
      <input
        id="title"
        placeholder="Task title"
        className="p-4 border border-black rounded text-xl"
        type="text"
      />
      <div
        id="datetime-wrapper"
        className="text-xl flex flex-row gap-4 items-center"
      >
        <label htmlFor="datetime">Due date/time: </label>
        <input
          id="datetime"
          type="datetime-local"
          className="border border-black rounded p-2"
        />
      </div>
      <textarea
        id="description"
        name="description"
        rows={3}
        placeholder="Task description"
        className="border border-black rounded p-4 resize-none"
      ></textarea>
      <div id="button-wrapper" className="w-[80%] ml-[20%] flex flex-row gap-4">
        <Button
          text="Cancel"
          textColor="text-redNCS"
          borderColor="border-redNCS"
          bgColor="bg-white"
          onClick={onCancel}
        />
        <Button
          text="Create new task"
          textColor="text-white"
          borderColor="border-redNCS"
          bgColor="bg-redNCS"
          type="submit"
        />
      </div>
    </form>
  );
};

export default NewTaskModalForm;
