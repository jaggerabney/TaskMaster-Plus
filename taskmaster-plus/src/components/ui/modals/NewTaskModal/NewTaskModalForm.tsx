"use client";

import React, { useContext, useRef, useState } from "react";

import Button from "../../Button";
import NewTaskRepeatForm from "./NewTaskRepeatForm/NewTaskRepeatForm";
import { ListContext, Task } from "@/contexts/ListContext";
import { buildRRuleStr, defaultRepeatFormState } from "@/utils/Util";

export interface NewTaskModalFormType {
  onSubmit: (task: Task) => void;
  onCancel: () => void;
}

const NewTaskModalForm: React.FC<NewTaskModalFormType> = ({
  onSubmit,
  onCancel
}) => {
  const [error, setError] = useState<string>("");
  const [repeatFormState, setRepeatFormState] = useState(
    defaultRepeatFormState
  );
  const listContext = useContext(ListContext);
  const listDropdownRef = useRef<HTMLOptionElement>(null);

  function formIsValid() {
    return error.length === 0;
  }

  function formChangeHandler() {
    setError("");
  }

  function formSubmitHandler(event: React.SyntheticEvent<HTMLFormElement>) {
    event.preventDefault();

    // the below line used to be:
    // const payload = event.target as typeof event.target & {
    // ...but was changed because extracting the values this way led to null values being
    // returned in the unit test that ensures you can't set a due date in the past.
    // isn't programming lovely?
    const payload = event.currentTarget
      .elements as typeof event.currentTarget.elements & {
      title: { value: string };
      date: { value: string };
      time: { value: string };
      description: { value: string };
    };

    const taskId = Math.random();
    const listId = Number(listDropdownRef.current?.id);
    const taskTitle = payload.title.value;
    let taskDueDate = undefined;
    let taskDescription = undefined;

    // If listId doesn't match any IDs in ListContext
    if (!listContext.state.lists.map((list) => list.id).includes(listId)) {
      return setError("Error: invalid list.");
    }

    // If title is empty
    if (!taskTitle) {
      return setError("Error: title is required.");
    }

    // If time is set and date is not
    if (payload.time.value && !payload.date.value) {
      return setError("Error: time cannot be set without date.");
    }

    if (payload.date.value) {
      const dateValues = payload.date.value.split("-");
      const year = Number(dateValues[0]);
      const month = Number(dateValues[1]) - 1; // months are 0-indexed in JavaScript :(
      const day = Number(dateValues[2]);
      const utcDueDate = new Date(year, month, day);

      if (utcDueDate.getTime() < Date.now()) {
        return setError("Error: due date cannot be in the past.");
      }

      taskDueDate = new Date(
        utcDueDate.getTime() + utcDueDate.getTimezoneOffset() * 60000
      );
    }

    if (payload.time.value) {
      // eventTarget.time.value format: HH:MM <- military time!
      // e.g., 20:20 for 8:20pm
      const hoursAndMinutes = payload.time.value.split(":");
      const hours = Number(hoursAndMinutes[0]);
      const minutes = Number(hoursAndMinutes[1]);

      taskDueDate?.setHours(hours, minutes);
    }

    if (payload.description.value) {
      taskDescription = payload.description.value;
    }

    if (formIsValid()) {
      const rruleStr: string = buildRRuleStr(repeatFormState);

      onSubmit({
        id: taskId,
        listId,
        title: taskTitle,
        completed: false,
        dueDate: taskDueDate,
        description: taskDescription,
        rrule: rruleStr
      });
    }
  }

  return (
    <form
      id="new-task-modal-form"
      onSubmit={formSubmitHandler}
      onChange={formChangeHandler}
      className="flex flex-col gap-4"
    >
      <div
        id="list-dropdown-wrapper"
        className="flex flex-row gap-4 items-center"
      >
        <label htmlFor="list" className="text-xl">
          List:
        </label>
        <select
          id="list"
          name="list"
          className="border border-black rounded text-xl p-2"
        >
          {listContext.state.lists.map((list) => (
            <option
              key={list.id}
              id={list.id.toString()}
              value={list.title}
              ref={listDropdownRef}
            >
              {list.title}
            </option>
          ))}
        </select>
      </div>
      <input
        id="title"
        placeholder="Task title"
        className="p-4 border border-black rounded text-xl"
        type="text"
      />
      <div
        id="datetime-wrapper"
        className="text-xl flex flex-row gap-4 items-center"
      >
        Due:
        <input
          id="date"
          type="date"
          className="border border-black rounded p-2"
        />
        <input
          id="time"
          type="time"
          className="border border-black rounded p-2"
        />
      </div>
      <textarea
        id="description"
        name="description"
        rows={3}
        placeholder="Task description"
        className="border border-black rounded p-4 resize-none"
      />
      <NewTaskRepeatForm data={repeatFormState} onChange={setRepeatFormState} />
      <div className="flex flex-row justify-center text-redNCS">
        {error.length > 0 && error}
      </div>
      <div id="button-wrapper" className="w-[80%] ml-[20%] flex flex-row gap-4">
        <Button
          text="Cancel"
          textColor="text-redNCS"
          borderColor="border-redNCS"
          bgColor="bg-white"
          onClick={onCancel}
        />
        <Button
          text="Create new task"
          textColor="text-white"
          borderColor="border-redNCS"
          bgColor="bg-redNCS"
          type="submit"
        />
      </div>
    </form>
  );
};

export default NewTaskModalForm;
