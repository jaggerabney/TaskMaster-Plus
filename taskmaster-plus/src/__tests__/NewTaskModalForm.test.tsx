import "@testing-library/jest-dom";
import { render, screen, fireEvent } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";

import NewTaskModalForm from "@/components/ui/modals/NewTaskModal/NewTaskModalForm";
import { ListContext, ListContextType } from "@/contexts/ListContext";

const customRender = (
  ui: React.JSX.Element,
  providerProps: ListContextType
) => {
  return render(
    <ListContext.Provider value={providerProps}>{ui}</ListContext.Provider>
  );
};

describe("New task modal form component", () => {
  const mockNewTaskModalForm = (
    <NewTaskModalForm onSubmit={() => {}} onCancel={() => {}} />
  );
  const defaultProps: ListContextType = {
    state: {
      lists: [
        {
          id: 1,
          title: "Test list 1",
          visible: true,
          tasks: []
        },
        {
          id: 2,
          title: "Test list 2",
          visible: true,
          tasks: []
        },
        {
          id: 3,
          title: "Test list 3",
          visible: true,
          tasks: []
        }
      ]
    },
    dispatch: () => {}
  };

  it("renders correctly", () => {
    const newTaskModalForm = render(mockNewTaskModalForm);

    expect(newTaskModalForm).toMatchSnapshot();
  });

  it("displays oldest list as default list option", () => {
    customRender(mockNewTaskModalForm, defaultProps);
    const listDropdown = screen.getByText(/test list 1/i);

    expect(listDropdown).not.toBeNull();
  });

  it("displays correct number of list options", () => {
    customRender(mockNewTaskModalForm, defaultProps);
    const listDropdownOptionElements = screen.getAllByRole("option");

    expect(listDropdownOptionElements.length).toBe(
      defaultProps.state.lists.length
    );
  });

  it("allows user to change list when there are multiple", () => {
    customRender(mockNewTaskModalForm, defaultProps);
    const optionToSelect = defaultProps.state.lists[2].title;

    userEvent.selectOptions(
      screen.getByRole("combobox"),
      screen.getByRole("option", { name: optionToSelect })
    );

    const listDropdown = screen.getByText(/test list 3/i);
    expect(listDropdown).not.toBeNull();
  });

  it("displays an error when no list is selected", () => {
    customRender(mockNewTaskModalForm, {
      state: { lists: [] },
      dispatch: () => {}
    });

    const createTaskButton = screen.getByText(/create new task/i);
    fireEvent.click(createTaskButton);

    const errorText = screen.queryByText(/error: invalid list/i);
    expect(errorText).not.toBeNull();
  });

  it('displays "Task title" placeholder text', () => {
    customRender(mockNewTaskModalForm, defaultProps);
    const taskTitleInput = screen.queryByPlaceholderText(/task title/i);

    expect(taskTitleInput).not.toBeNull();
  });

  it("allows user to enter text into title input", async () => {
    customRender(mockNewTaskModalForm, defaultProps);
    const taskTitleInput = screen.getByPlaceholderText(/task title/i);

    await userEvent.type(taskTitleInput, "lorem ipsum");

    const taskTitle = screen.queryByDisplayValue(/lorem ipsum/i);
    expect(taskTitle).not.toBeNull();
  });

  it("displays an error when no title is specified", async () => {
    customRender(mockNewTaskModalForm, defaultProps);
    const createNewTaskButton = screen.getByText(/create new task/i);

    await userEvent.click(createNewTaskButton);

    const errorText = screen.queryByText(/error: title is required/i);
    expect(errorText).not.toBeNull();
  });

  it("allows user to change due date", () => {
    customRender(mockNewTaskModalForm, defaultProps);
    const taskDueDateInput = document.getElementById(
      "date"
    ) as HTMLInputElement;

    fireEvent.change(taskDueDateInput, { target: { value: "2030-01-01" } });

    expect(taskDueDateInput.value).toMatch(/2030-01-01/i);
  });

  it("allows for task submission without optional values", async () => {
    const mockOnSubmit = jest.fn();
    customRender(mockNewTaskModalForm, defaultProps);
    const taskTitleInput: HTMLInputElement =
      screen.getByPlaceholderText(/task title/i);
    const newTaskModalForm = document.getElementById(
      "new-task-modal-form"
    ) as HTMLFormElement;
    newTaskModalForm.onsubmit = mockOnSubmit();

    await userEvent.type(taskTitleInput, "test task");
    fireEvent.submit(newTaskModalForm);

    expect(mockOnSubmit).toHaveBeenCalled();
  });

  // this unit test put me out of the mood to code for about three days...
  // the way i was extracting values from the form - specifically, by using event.target instead of
  // event.currentTarget.elements in the submit handler - led to null values being passed to state,
  // even though it hasn't caused problems in any of the above tests.
  // thus, the rendered error text was the no title one, as the  it's the top one in the file.
  it("doesn't allow user to assign a due date in the past", async () => {
    customRender(mockNewTaskModalForm, defaultProps);
    const taskTitleInput: HTMLInputElement =
      screen.getByPlaceholderText(/task title/i);
    const taskDueDateInput = document.getElementById(
      "date"
    ) as HTMLInputElement;
    const createNewTaskButton = screen.getByText(/create new task/i);

    await userEvent.type(taskTitleInput, "test task");
    fireEvent.change(taskDueDateInput, { target: { value: "2002-05-11" } });
    await userEvent.click(createNewTaskButton);

    const errorText = screen.queryByText(
      /error: due date cannot be in the past./i
    );
    expect(errorText).not.toBeNull();
  });

  it("allows user to change the due time", async () => {
    customRender(mockNewTaskModalForm, defaultProps);
    const taskDueTimeInput = document.getElementById(
      "time"
    ) as HTMLInputElement;

    await userEvent.type(taskDueTimeInput, "09:20", {
      initialSelectionStart: 0,
      initialSelectionEnd: taskDueTimeInput.value.length
    });

    expect(taskDueTimeInput.value).toMatch(/09:20/i);
  });

  it("doesn't allow user to set a due time with a corresponding date", async () => {
    customRender(mockNewTaskModalForm, defaultProps);
    const taskTitleInput: HTMLInputElement =
      screen.getByPlaceholderText(/task title/i);
    const taskDueTimeInput = document.getElementById(
      "time"
    ) as HTMLInputElement;
    const createNewTaskButton = screen.getByText(/create new task/i);

    await userEvent.type(taskTitleInput, "test task");
    await userEvent.type(taskDueTimeInput, "09:20", {
      initialSelectionStart: 0,
      initialSelectionEnd: taskDueTimeInput.value.length
    });
    await userEvent.click(createNewTaskButton);

    const errorText = screen.queryByText(
      /error: time cannot be set without date./i
    );
    expect(errorText).not.toBeNull();
  });

  it("should allow the user to change the description", async () => {
    customRender(mockNewTaskModalForm, defaultProps);
    const taskDescriptionInput: HTMLTextAreaElement =
      screen.getByPlaceholderText(/task description/i);

    await userEvent.type(taskDescriptionInput, "test description");

    expect(taskDescriptionInput.value).toMatch(/test description/i);
  });
});
