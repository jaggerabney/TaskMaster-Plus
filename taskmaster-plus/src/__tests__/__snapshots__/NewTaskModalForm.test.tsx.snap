// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New task modal form component renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="flex flex-col gap-4"
        id="new-task-modal-form"
      >
        <div
          class="flex flex-row gap-4 items-center"
          id="list-dropdown-wrapper"
        >
          <label
            class="text-xl"
            for="list"
          >
            List:
          </label>
          <select
            class="border border-black rounded text-xl p-2"
            id="list"
            name="list"
          />
        </div>
        <input
          class="p-4 border border-black rounded text-xl"
          id="title"
          placeholder="Task title"
          type="text"
        />
        <div
          class="text-xl flex flex-row gap-4 items-center"
          id="datetime-wrapper"
        >
          Due:
          <input
            class="border border-black rounded p-2"
            id="date"
            type="date"
          />
          <input
            class="border border-black rounded p-2"
            id="time"
            type="time"
          />
        </div>
        <textarea
          class="border border-black rounded p-4 resize-none"
          id="description"
          name="description"
          placeholder="Task description"
          rows="3"
        />
        <div
          class="text-xl flex flex-row gap-4 items-center"
          id="repeat-checkbox-wrapper"
        >
          <input
            id="repeat-checkbox"
            type="checkbox"
          />
          <label
            for="repeat-checkbox"
          >
            Repeat
          </label>
        </div>
        <div
          class="flex flex-row justify-center text-redNCS"
        />
        <div
          class="w-[80%] ml-[20%] flex flex-row gap-4"
          id="button-wrapper"
        >
          <button
            class="px-4 py-2 w-full text-redNCS border-redNCS bg-white whitespace-nowrap select-none font-bold border-2 rounded hover:scale-[1.03] duration-100"
          >
            Cancel
          </button>
          <button
            class="px-4 py-2 w-full text-white border-redNCS bg-redNCS whitespace-nowrap select-none font-bold border-2 rounded hover:scale-[1.03] duration-100"
            type="submit"
          >
            Create new task
          </button>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="flex flex-col gap-4"
      id="new-task-modal-form"
    >
      <div
        class="flex flex-row gap-4 items-center"
        id="list-dropdown-wrapper"
      >
        <label
          class="text-xl"
          for="list"
        >
          List:
        </label>
        <select
          class="border border-black rounded text-xl p-2"
          id="list"
          name="list"
        />
      </div>
      <input
        class="p-4 border border-black rounded text-xl"
        id="title"
        placeholder="Task title"
        type="text"
      />
      <div
        class="text-xl flex flex-row gap-4 items-center"
        id="datetime-wrapper"
      >
        Due:
        <input
          class="border border-black rounded p-2"
          id="date"
          type="date"
        />
        <input
          class="border border-black rounded p-2"
          id="time"
          type="time"
        />
      </div>
      <textarea
        class="border border-black rounded p-4 resize-none"
        id="description"
        name="description"
        placeholder="Task description"
        rows="3"
      />
      <div
        class="text-xl flex flex-row gap-4 items-center"
        id="repeat-checkbox-wrapper"
      >
        <input
          id="repeat-checkbox"
          type="checkbox"
        />
        <label
          for="repeat-checkbox"
        >
          Repeat
        </label>
      </div>
      <div
        class="flex flex-row justify-center text-redNCS"
      />
      <div
        class="w-[80%] ml-[20%] flex flex-row gap-4"
        id="button-wrapper"
      >
        <button
          class="px-4 py-2 w-full text-redNCS border-redNCS bg-white whitespace-nowrap select-none font-bold border-2 rounded hover:scale-[1.03] duration-100"
        >
          Cancel
        </button>
        <button
          class="px-4 py-2 w-full text-white border-redNCS bg-redNCS whitespace-nowrap select-none font-bold border-2 rounded hover:scale-[1.03] duration-100"
          type="submit"
        >
          Create new task
        </button>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
